### ISSUE TITLE
Classpath for compiler interferes with classpath of what is being compiled

### ISSUE THREAD

**Issue by hakanai** [tbroyer/gradle-errorprone-plugin-v0.0.x#67]
Error Prone bumped their dependency on Guava to version 23.5. Our own project was asking for version 23.0, so when the new version of Error Prone came out a couple days ago, our compiles started failing.

The workaround we had to take was to bump our own dependency on Guava up to 23.5, which then pulled in checker-qual and caused some other build issues which we're still working through.

I don't think that the classpath of the compiler is supposed to cause version conflicts with the classes being compiled, though - any dependencies it has should have been isolated from ours.

Demonstration project: https://github.com/trejkaz/gradle-errorprone-classpathleak
(Blows up on a different jar - I guess in the case of JSR305 the project where we discovered the issue must have already been forcing a later version for some reason.)

```
$ ./gradlew compileJava
> Task :compileJava FAILED

FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':compileJava'.
> Could not resolve all dependencies for configuration ':errorprone'.
   > A conflict was found between the following modules:
      - com.google.code.findbugs:jsr305:3.0.0
      - com.google.code.findbugs:jsr305:1.3.9

* Try:
Run with --stacktrace option to get the stack trace. Run with --info or --debug option to get more log output. Run with --scan to get full insights.

* Get more help at https://help.gradle.org

BUILD FAILED in 0s
1 actionable task: 1 executed
```

**Comment by tbroyer** [comment]
AFAICT from your demo project, this has nothing to do with the Error Prone classpath "leaking" to the compile classpath.

If you comment out the `failOnVersionConflict` resolution strategy and run `./gradlew dependencies --configuration=errorprone`, you see that JSR305, checker-qual and error_prone_annotations all come in different versions from Guava (as a transitive dependency of Error Prone) and Error Prone itself.
```
$ ./gradlew dependencies --configuration=errorprone

> Task :dependencies

------------------------------------------------------------
Root project
------------------------------------------------------------

errorprone
\--- com.google.errorprone:error_prone_core:latest.release -> 2.3.1
     +--- com.google.errorprone:error_prone_annotation:2.3.1
     |    \--- com.google.guava:guava:23.5-jre
     |         +--- com.google.code.findbugs:jsr305:1.3.9 -> 3.0.0
     |         +--- org.checkerframework:checker-qual:2.0.0 -> 2.5.0
     |         +--- com.google.errorprone:error_prone_annotations:2.0.18 -> 2.3.1
     |         +--- com.google.j2objc:j2objc-annotations:1.1
     |         \--- org.codehaus.mojo:animal-sniffer-annotations:1.14
     +--- com.google.errorprone:error_prone_type_annotations:2.3.1
     +--- com.google.errorprone:error_prone_check_api:2.3.1
     |    +--- com.google.errorprone:error_prone_annotation:2.3.1 (*)
     |    +--- com.google.code.findbugs:jsr305:3.0.0
     |    +--- org.checkerframework:dataflow:2.5.0
     |    |    +--- org.checkerframework:checker-qual:2.5.0
     |    |    \--- org.checkerframework:javacutil:2.5.0
     |    |         \--- org.checkerframework:checker-qual:2.5.0
     |    +--- com.google.errorprone:javac:9+181-r4173-1
     |    +--- com.googlecode.java-diff-utils:diffutils:1.3.0
     |    +--- com.google.errorprone:error_prone_annotations:2.3.1
     |    \--- com.github.kevinstern:software-and-algorithms:1.0
     +--- com.github.stephenc.jcip:jcip-annotations:1.0-1
     +--- org.pcollections:pcollections:2.1.2
     +--- com.google.guava:guava:23.5-jre (*)
     +--- com.google.auto:auto-common:0.10
     |    \--- com.google.guava:guava:23.5-jre (*)
     +--- com.google.code.findbugs:jFormatString:3.0.0
     +--- com.google.code.findbugs:jsr305:3.0.0
     +--- org.checkerframework:dataflow:2.5.0 (*)
     +--- com.google.errorprone:javac:9+181-r4173-1
     \--- com.google.errorprone:error_prone_annotations:2.3.1

(*) - dependencies omitted (listed previously)

A web-based, searchable dependency report is available by adding the --scan option.

BUILD SUCCESSFUL in 0s
1 actionable task: 1 executed
```

You'll want to change your build to:
 * pin a version of Error Prone so that at least you don't run into issues due to a new version of Error Prone being published and the `latest.release` version being resolved. This is not related to your issue but is a good practice.
 * then either:
   * disable the `failOnVersionConflict` on the `errorprone` configuration, e.g.

        ```gradle
        configurations {
          all {
            if (it.name == "errorprone") return;
            resolutionStrategy {
              failOnVersionConflict()
            }
          }
        }
        ```

   * or "fix" the conflicts using Gradle resolution rules for the `errorprone` configuration:

        ```gradle
        configurations {
          errorprone {
            resolutionStrategy {
              force "com.google.code.findbugs:jsr305:3.0.0"
              force "com.google.errorprone:error_prone_annotations:2.3.1"
              force "org.checkerframework:checker-qual:2.5.0"
            }
          }
          all {
            resolutionStrategy {
              failOnVersionConflict()
            }
          }
        }
        ```

**Comment by hakanai** [comment]
That conflict shouldn't even exist though. It only exists because the dependency on a different version of jsr305 is leaking from the error-prone classpath into the classpath of what it's compiling.

I already know how to work around the problem, but I thought I would report it anyway.

**Comment by tbroyer** [comment]
> That conflict shouldn't even exist though. It only exists because the dependency on a different version of jsr305 is leaking from the error-prone classpath into the classpath of what it's compil

No, this really has nothing to do with any "leak". You can reproduce the error with the following build script (a single configuration so no possible "leak", no error prone plugin) and `./gradlew dependencies`:
```gradle
plugins {
  id "base"
}
repositories {
  mavenCentral()
}
dependencies {
  add("default", "com.google.errorprone:error_prone_core:2.3.1")
}
configurations {
  all {
    resolutionStrategy {
      failOnVersionConflict()
    }
  }
}
```

**Comment by hakanai** [comment]
But that doesn't explain why we get a clash between Guava 23.0 and Guava 23.5, when our own project only requires 23.0.

Are you saying that errorprone also depends on two versions of Guava? If so, why the hell?

**Comment by tbroyer** [comment]
That's right. Then there must be something else that your demo project doesn't demonstrate.

I'm certain this is not a "classpath leak" and/or due to this plugin, but if you can update your demo project to actually reproduce what you're experiencing I can try to help you pinpoint the issue.

**Comment by tbroyer** [comment]
Or could you maybe share the results of a `./gradlew dependencies` after removing the resolution strategies that make it fail? Or try using `./gradlew dependencyInsight …`

**Comment by hakanai** [comment]
I've done the research on this now, essentially paring down the real project until I found the bit which breaks it, and it turns out to essentially be this:

```
  configurations {
    all {
      resolutionStrategy {
        force 'com.google.guava:guava:23.0'
        failOnVersionConflict()
      }
    }
```

Since we have `configurations.all` here, I guess this messes with the classpath errorprone wants to run in, so we should have something else here. I'm just not quite sure what that something else would be, but something which matches all compile and runtime configurations without matching other tools' configurations.

It sucks that we even need to use `force` but we only have to force versions because of `failOnVersionConflict()`. I wish we didn't have to fail on version conflict, but because Gradle sucks at resolving dependencies we don't appear to have any other option.

**Comment by tbroyer** [comment]
> Since we have configurations.all here, I guess this messes with the classpath errorprone wants to run in, so we should have something else here. I'm just not quite sure what that something else would be, but something which matches all compile and runtime configurations without matching other tools' configurations.

See above, you could _blacklist_ the `errorprone` configuration using:
```gradle
configurations {
  all {
    if (it.name == "errorprone") return;
    resolutionStrategy {
      force 'com.google.guava:guava:23.0'
      failOnVersionConflict()
    }
  }
}
```
(or include the `resolutionStrategy` block inside an `if (it.name != "errorprone")`)

or whitelist the configurations you apply the resolution strategy to, e.g. based on the source sets:
```gradle
sourceSets.all {
  [it.compileClasspathConfigurationName, it.annotationProcessorConfigurationName, it.runtimeClasspathConfigurationName].each {
    configurations[it].resolutionStrategy {
      …
    }
  }
}
```
(if you apply this in an `allprojects` or `subprojects` block in the root project of a multi-project build, then include the block in a `plugins.withType(JavaBasePlugin) { … }` to make sure the project will have a `sourceSets` property)

> I wish we didn't have to fail on version conflict, but because Gradle sucks at resolving dependencies we don't appear to have any other option.

This is the first time I hear about this, and would like to know more!

**Comment by hakanai** [comment]
The original trigger, which was a fair while ago now, was a library clash where one copy was v1.0 and the other copy was v2.0, which broke compilation. This then led to investigating other clashes and we found a lot more of them where the issues had only occurred at runtime. We eventually decided to force a failure in this situation so that we could make a proper decision each time.

But I think the core to the problem is actually a complex mess of responsibility / blame:

* Maven supports a bare version in the first place, when it should have supported only ranges. I'm assuming that ranges were bolted on later, perhaps in response to OSGi (the formats are the same), and the community never adopted ranges as the default way to declare dependencies.
* Maven says that if a bare version is specified, it's equivalent to "this version or greater".
* Gradle implements the Maven behaviour, so if 1.0 and 2.0 both occur, everyone ends up using 2.0 and you get this kind of clash. (If you're lucky, compilation breaks, so you get to find out.)
* Java library developers in general are either too lazy to declare the actual version range of dependencies they support, or simply unaware that it's possible to do so. OSGi developers seem to pay close attention to it, but people who aren't aware of OSGi will tend to declare a hard dependency on the one version of a library they test against.
* Gradle bloggers have spotted seen numerous times _actively discouraging_ setting proper version ranges, leading to even less people correctly declaring their supported dependencies.

At this point I think it's basically so far gone that the only way to get out of the mess is to abandon the Maven model of dependencies and make something closer to OSGi. I don't know if Jigsaw has brought us another alternative, though, as we have to wait for 100 or so libraries to add Java 9 support before we can consider moving ourselves.
